### ISSUE TITLE
strictNullChecks does not detect assertions done before property access

### ISSUE THREAD

**Issue by connium** [Microsoft/TypeScript#13652]
**TypeScript Version:**  2.1.5

**Code**
In my jasmine tests I use `.toBeDefined(...)` to check whether the tested value is filled properly.
```ts
interface Person {
  name: string;
  address?: Address;
}

interface PersonService {
  getByName(name: string): Person;
}

it('should work as expected', () => {
  const person = personService.getByName('Norris');

  expect(person.address).toBeDefined();
  expect(person.address.state).toEqual('Oklahoma');
});
```
Running the compiler with `strictNullChecks` enabled results in a `error TS2532: Object is possibly 'undefined'.` on the line where `person.address.state` is accessed.

Checking explicitly for `undefined` resolves this issue:
```ts
it('should work as expected', () => {
  const person = personService.getByName('Norris');

  if (person.address === undefined) {
    throw new Error('address is undefined');
  }
  expect(person.address.state).toEqual('Oklahoma');
});
```

As this is a common pattern in my code, I thought it would be a good idea to extract the check in order to keep the tests clean and simple.
```ts
function assertNotUndefined(value: any): void|never {
  if (value === undefined) {
    throw new Error('value is undefined');
  }
}

it('should work as expected', () => {
  const person = personService.getByName('Norris');

  assertNotUndefined(person.address);
  expect(person.address.state).toEqual('Oklahoma');
});
```
Now I get the compiler error again.

**Expected behavior:**
The compiler does not complain about access to an potentially undefined variable when `--strictNullChecks` is set.

**Actual behavior:**
The compiler says `error TS2532: Object is possibly 'undefined'.`

**Comment by RyanCavanaugh** [comment]
See  #9693

A pattern you can use is:
```ts
function assertNotUndefined<T>(value: T | undefined): value is T{
  if (value === undefined) {
    throw new Error('value is undefined');
    }
  return true;
}

it('should work as expected', () => {
  const person: {address?: {state: string}} = personService.getByName('Norris');

  if (!assertNotUndefined(person.address)) return;
  expect(person.address.state).toEqual('Oklahoma');
});
```

**Comment by connium** [comment]
You are right, my issue is similar to the one described in #9693. Didn't find it when looking for solutions to my problem somehow.

I like the idea of type guards as they are proposed in #8655. So closing this in favor of #8655.
